### ISSUE TITLE
Efficient reduceSum over entire tensor

### ISSUE THREAD

**Issue by johnynek** [tensorflow/java#368]
In python, you can do `tf.reduce_sum(x)` to reduce all values down to a single scalar. see python docs:
https://www.tensorflow.org/api_docs/python/tf/math/reduce_sum

but here: https://github.com/tensorflow/java/blob/4c1c2718ef1c4c386891d92cf0d774d7e77a7dc1/tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/op/core/ReduceSum.java#L73

it seems we can only do a single axis at a time.

I can certainly implement a library function to handle the case without an axis by looping over all axes, but I wonder if tensorflow internally can do it more efficiently without having to do one per axis (which will generally have a couple of smallish dimensions and those calls seem somewhat wasteful).

**Comment by rnett** [comment]
`axis` does not have to be a scalar.  It's a bit confusing, since there's no indication of what shape axis should be, but I suspect it could be a 1d array, say `[0 .. shape.numDimensions]` to reduce the whole thing.  Certainly not a friendly API though.

**Comment by Craigacp** [comment]
Axis can be a tensor representing all of the inputs, so you should be able to use a single `ReduceSum` to do the full reduction.

**Comment by johnynek** [comment]
oh, great, thanks!
